### PR TITLE
Reducing the GEMDigiSimLink size and turning off gem multi bx readout during Run3

### PR DESCRIPTION
**Pull request by yeckang**
#### PR description:

* Reducing the size of the GEMDigiSimLinks
* Because the GEM is not supporting multi bx readout during Run 3, we made a bx filter in GEMDigitizer.

#### PR validation:

* The branch has tested with the following commands to check if it works well in Run 3 and Phase 2 conditions
```
runTheMatrix.py -w upgrade -l 11611.0
runTheMatrix.py -w upgrade -l 35811.0
```
* The code quality has checked with following commands.
```
scram build code-check
scram build code-format
scram build -k -j 16 check-headers
```

### CODE REVIEW

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

I would guess that problems in testing are not connected with this PR.

File: SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py [hunk 1]
```diff
@@ -7,10 +7,10 @@
 premix_stage2.toModify(simMuonGEMDigis, mixLabel = "mixData")
 
 from Configuration.Eras.Modifier_run2_common_cff import run2_common
-run2_common.toModify( simMuonGEMDigis, instLumi = 1.5)
+run2_common.toModify( simMuonGEMDigis, instLumi = 1.5, bx0filter = True)
```
Comment (srimanob) [inline SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py:10]: Should you keep it to False, so Run-2 behaviour will not be changed?

Comment (yeckang) [reply SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py:10]: There were only 5 GE1/1 super chambers for the slice test.
And as far as I know, they didn't support the multi bx mode at that moment.
@jshlee Could you confirm it?

**Comment by jfernan2** [comment]
@yeckang there are changes in occupancy plots for the Validation folder
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_1_X_2021-08-12-1100+8eed5a/44744/dqm-histo-comparison-summary.html
Can you confirm they are expected?

File: SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py (hunk 1 above)
Comment (jshlee) [reply SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py:10]: bx0filter = True is correct for run2

**Comment by yeckang** [comment]
@jfernan2 
In principle, it will remove the non-0 bx digis from Run2 and Run3 scenarios.
So the occupancy plots could be changed.
But the changes in the links are in Phase 2 scenarios.
So I tested the runTheMatrix.py with 28211.0, which has the same global tag and geometry with 28234.0, again. And I saw the non-0 bx digis are in the output.
I guess the differences are coming from the random number that we use to applying the efficiency to make digis.

**Comment by jfernan2** [comment]
@yeckang can you elaborate about those random numbers? The randomness should be identical in the Jenins tests performed here in github

**Comment by jfernan2** [comment]
Have you seen these changes in your private validation for this PR?

**Comment by jfernan2** [comment]
What worries me the most is the migration from p, pi+, pi-, K...to other in the last bien for all these plots: https://tinyurl.com/ydpyfszu

File: SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py (hunk 1 above)
Comment (srimanob) [reply SimMuon/GEMDigitizer/python/muonGEMDigis_cfi.py:10]: OK, from https://github.com/cms-sw/cmssw/pull/34840#issuecomment-899366332 you expect to change the behaviour of Run-2.

**Comment by yeckang** [comment]
I see. The most of particles were moved into the last bin.
I just checked my (private) output. But I don't see such a tendency.
The content of the last bin is usually zero. It's something strange.

**Comment by yeckang** [comment]
Sorry. What I checked is the GEMDigisV, not GEMRecHitsV. I see the same problem from my private output.
It should be something wrong. Let me check.

**Comment by yeckang** [comment]
I found the reason.
https://github.com/yeckang/cmssw/blob/gemDigiUpdate_12_1_0_pre1/SimDataFormats/GEMDigiSimLink/interface/GEMDigiSimLink.h#L33
The `int8_t` is not big enough to contain the pdg id.
I will change its type to `int16_t` and check my private output again. 
After I will push it again.
Thanks for pointing the error.

**Comment by srimanob** [comment]
@cmsbuild please test

**Comment by jfernan2** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
It is not originated by this PR, but the Static Analyzer point out to a bug in `GEMDigiMatcher::matchClustersToSimTrack`.

At [line 218](https://github.com/cms-sw/cmssw/blob/beb3c37912686e70980b3449b5401d3ec4193803/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc#L218) the logical local variable `isMatched` is not initialized to `false` (as it is done, for example, in the similar structure in `GEMDigiMatcher::matchDigisToSimTrack`). This leads to a possible undefined behaviour if the check that starts at [line 232](https://github.com/cms-sw/cmssw/blob/beb3c37912686e70980b3449b5401d3ec4193803/Validation/MuonGEMDigis/src/GEMDigiMatcher.cc#L232) is unsuccesfull.

Please profit of this PR and fix it.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by civanch** [comment]
+1

**Comment by jfernan2** [comment]
+1

**Pull request merged**